### PR TITLE
Add label selector option for Secret and ConfigMap sources

### LINKED ISSUE #256: Use label selector to add sources to a bundle
The `Bundle` CR allows to specify a list of `Secret` and `ConfigMap` by name in the `spec.sources`. In environments where the list of certificates to add to the `Bundle` is dynamic, it can become complicated to maintain the `Bundle`.

Would it be possible to specify, in a `Bundle` spec, a label selector for `Secret` and `ConfigMap`? That way the `Bundle` could be automatically updated, based on labels on `Secret` and `ConfigMap`, when those resources are created, updated and deleted.

### PR DESCRIPTION
**Pull request by ocampeau**
This PR adds support to use label selector for Secret and ConfigMap sources. 

Given a `Bundle` that looks like this:

```yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: some-bundle
spec:
  sources:
  - useDefaultCAs: true
  - configMapLabelSelector:
      key: ca.crt
      selector:
        matchLabels:
          fruit: apple
  - secretMapLabelSelector:
      key: ca.crt
      selector:
        matchLabels:
          fruit: banana
  target:
    configMap:
      key: ca.crt
```
this behavior will happen:

* all `Secret` that has label `fruit: banana` and a key `ca.crt` will have the value of `ca.crt` included in the `Bundle`
* all `ConfigMap` that has label `fruit: apple` and a key `ca.crt` will have the value of `ca.crt` included in the `Bundle`

Multiple `configMapLabelSelector` and `secretMapLabelSelector` are supported.

In a case like this:
```yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: some-bundle
spec:
  sources:
  - useDefaultCAs: true
  - configMapLabelSelector:
      key: ca.crt
      selector:
        matchLabels:
          fruit: apple
  - configMapLabelSelector:
      key: ca.crt
      selector:
        matchLabels:
          fruit: banana
  target:
    configMap:
      key: ca.crt
```

then:

* all `ConfigMap` that has label `fruit: banana` and a key `ca.crt` will have the value of `ca.crt` included in the `Bundle`
* all `ConfigMap` that has label `fruit: apple` and a key `ca.crt` will have the value of `ca.crt` included in the `Bundle`

fixes #256

### CODE REVIEW

**Comment by erikgb** [comment]
I wonder if we need to extend the API with "new source types"; could we instead just extend the existing source `configMap` and `secret` types to allow **either** `name` **or** `selector`? Making the following bundle valid:

````yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: some-bundle
spec:
  sources:
  - useDefaultCAs: true
  - configMap:
      key: ca.crt
      name: my-configmap
  - configMap:
      key: ca.crt
      selector:
        matchLabels:
          fruit: apple
  - secret:
      key: ca.crt
      name: my-secret
  - secret:
      key: ca.crt
      selector:
        matchLabels:
          fruit: banana
  target:
    configMap:
      key: ca.crt
````

**Comment by ocampeau** [comment]
Yeah definitely! I will update my PR to reuse the existing source type like you mention. It's much better like that.

**Comment by ocampeau** [comment]
It's done. Let me know what you think.

File: pkg/webhook/validation.go
Comment (erikgb) [inline pkg/webhook/validation.go:105]: We must also validate that the user hasn't set both `name` and `selector`.

Comment (ocampeau) [reply pkg/webhook/validation.go:105]: I just added a test for that. Let me know what you think!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:70]:
```suggestion
	// list of ConfigMap's `data` key using label selector, in the trust Namespace.
```

For consistency.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:75]:
```suggestion
	// list of Secret's `data` key using label selector, in the trust Namespace.
```

For consistency.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:163]:
```suggestion
	Name string `json:"name",omitempty`
```

File: pkg/bundle/sync.go
Comment (erikgb) [inline pkg/bundle/sync.go:26]: This import should be moved to the preceding block. This block is for stdlib imports.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:168]: I notice that we have created a custom type for the namespace selector in this project API: `NamespaceSelector`. I wonder why? Maybe we want to own the API? @inteon WDYT? I am biased here, but leaning towards using the upstream K8s type.

File: pkg/bundle/sync.go
Comment (erikgb) [inline pkg/bundle/sync.go:162]: I think we should rather use `metav1.LabelSelectorAsSelector` here - to get better/safer error handling. The function `metav1.FormatLabelSelector` seems to return "funny" strings in case of error. Maybe create a small helper to reduce duplication even more?

File: pkg/webhook/validation.go
Comment (erikgb) [inline pkg/webhook/validation.go:109]:
```suggestion
				el = append(el, field.Invalid(path, configMap, "source configMap name and label selector are both set; only one must be set"))
```

I think you need to take a look at the error message here. Not sure what would be most user-friendly message. Maybe take a look at the `oneOf` validation? This validation is what we need here, and is supported på OpenAPI, but not yet by controller-runtime.

File: pkg/webhook/validation.go
Comment (erikgb) [reply pkg/webhook/validation.go:105]: Looks like a good start, but a bit more work is needed. We could see if there is a way to patch the CRD with a [oneOf](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/), which is supported by OpenAPI/Kubernetes, but not yet by c/r, ref. https://github.com/kubernetes-sigs/controller-tools/issues/461.

**Review by erikgb — CHANGES_REQUESTED** [review]
Added some review comments, but mostly nits. I am prepared to move forward with this PR when the code looks good. Well done @ocampeau! 😍

File: pkg/bundle/sync.go
Comment (ocampeau) [reply pkg/bundle/sync.go]: Done!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (inteon) [reply pkg/apis/trust/v1alpha1/types_bundle.go:168]: I think it is very valid to embed the metav1 struct here.
Also, the LabelSelector seems to have good comments that can be used to generate useful OpenAPI descriptions.

File: pkg/bundle/sync.go
Comment (ocampeau) [reply pkg/bundle/sync.go]: I refactored this part a bit to reduce duplication and used `metav1.LabelSelectorAsSelector`. Let me know what you think!

File: pkg/webhook/validation.go
Comment (ocampeau) [reply pkg/webhook/validation.go]: When I add the `oneOf` validation manually in the CRD spec, this is the error message that is returned:

```bash
The Bundle "test" is invalid:
* <nil>: Invalid value: "": "spec.sources[1].secret" must validate one and only one schema (oneOf). Found none valid
* spec.sources[1].secret.name: Required value
* spec.sources[1].secret.selector: Required value
* <nil>: Invalid value: "": "spec.sources[2].configMap" must validate one and only one schema (oneOf). Found none valid
* spec.sources[2].configMap.name: Required value
* spec.sources[2].configMap.selector: Required value
```

Unless you says otherwise, I think trying to modify the CRD afterwards to inject the `oneOf` validation can be a fragile solution. So instead, since controller-runtime does not support `oneOf` OpenAPI validation out of the box, I tried to return an error message as close as possible to what the OpenAPI spec returns, with some minor deviations to be as user-friendly as possible.

It goes like that:

```bash
Error from server (Forbidden): error when creating "bundle.yaml": admission webhook "trust.cert-manager.io" denied the request: 
[
spec.sources.[2].configMap: Invalid value: "name: ' ', selector: nil": must validate one and only one schema (oneOf): [name, selector]. Found none valid,
spec.sources.[1].secret: Invalid value: "name: allo, selector: {}": must validate one and only one schema (oneOf): [name, selector]. Found none valid
]
```

1. Let's look at the first error, when both `name` and `selector` are not set:

```text
spec.sources.[2].configMap: Invalid value: "name: ' ', selector: nil": must validate one and only one schema (oneOf): [name, selector]. Found none valid
```

here, I simply hardcoded in the error message an empty string as the value of `name` and `nil` as the value of `selector`.

2. Let's look at the second error, when both `name` and `selector` are set:

```text
spec.sources.[1].secret: Invalid value: "name: allo, selector: {}": must validate one and only one schema (oneOf): [name, selector]. Found none valid
```

here, I parametrized the value of the `name` fields using the value provided by the user, but instead of parametrizing as well the value of `selector`, which turns out to be an ugly stringified version of a Golang map, I instead hardcoded `{}` to indicate that the value is not nil, it is set to some object.


Let me know what you think

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (ocampeau) [reply pkg/apis/trust/v1alpha1/types_bundle.go:168]: So I guess I can resolve this comment?

File: pkg/webhook/validation.go
Comment (ocampeau) [reply pkg/webhook/validation.go:105]: From what I have found Googling around a bit, most people seems to manually add that validation into a webhook instead of patching the CRD afterwards (because it gets overwritten every time we re-generate the CRDs).

See my other comments on the other threads about this issue: https://github.com/cert-manager/trust-manager/pull/258#discussion_r1445222407

Comment (inteon) [reply pkg/webhook/validation.go:105]: We already have a webhook, so I think there is no need for a oneOf OpenAPI validation.

File: pkg/bundle/controller.go
Comment (inteon) [inline pkg/bundle/controller.go:152]:
```suggestion
					} else {
					  if source.ConfigMap.Name == obj.GetName() {
						  return true
					  }
					}
```

File: pkg/bundle/controller.go
Comment (inteon) [inline pkg/bundle/controller.go:233]: I think you can use `metav1.LabelSelectorAsSelector` here.

**Comment by inteon** [comment]
@ocampeau This is an awesome improvement ❤️!
Thanks a lot for contributing this.
/approve
Can be LGTM'd once the two comments above have been resolved ^^

File: pkg/bundle/controller.go
Comment (ocampeau) [reply pkg/bundle/controller.go]: Indeed! Good catch!

**Comment by ocampeau** [comment]
I have resolved all conversation except this one: https://github.com/cert-manager/trust-manager/pull/258#discussion_r1445222407

Let me know if this is acceptable or not. 

Thanks a lot for accepting this PR by the way, it will really help us. It was a pleasure to contribute to this project.

File: pkg/webhook/validation.go
Comment (erikgb) [reply pkg/webhook/validation.go]: Thanks for the detailed work on this, @ocampeau! As long as the field reference is correct (example: `spec.sources.[1].secret`) , I wouldn't worry too much about the following text. I like your suggestion of dropping the default string formatting of a Golang struct.

Comment (ocampeau) [reply pkg/webhook/validation.go]: Cool! Thanks I am glad you like it.

**Comment by erikgb** [comment]
> Any ETA on when this could potentially be merged? Do you have documentation on the release velocity of this project?

Sorry, I have been very busy lately. Will try to review this weekend if @inteon hasn't reviewed and merged it before.

We have a couple of unreleased changes already, so I would suggest considering releasing 0.8.0 after this PR is merged. WDYT, @inteon? Sorry for the delay, @ocampeau!

File: pkg/bundle/controller.go
Comment (erikgb) [inline pkg/bundle/controller.go:22]: Nit: This should be moved to the next import block. This block is for stdlib imports.

File: pkg/bundle/controller.go
Comment (erikgb) [inline pkg/bundle/controller.go:226]:
```suggestion
// labelsMatchSelector returns true if objLabels matches the label selector
```

File: pkg/webhook/validation.go
Comment (erikgb) [inline pkg/webhook/validation.go:109]:
```suggestion
				el = append(el, field.Invalid(path, fmt.Sprintf("name: %s, selector: {}", configMap.Name), "must validate one and only one schema (oneOf): [name, selector]. Found both set"))
```

I am not sure about the error message here, but it should indicate that we detected both set - which is invalid.

**Review by erikgb — CHANGES_REQUESTED** [review]
Some minor nits left, but this is close to a LGTM from me!

File: pkg/bundle/controller.go
Comment (ocampeau) [reply pkg/bundle/controller.go:22]: I think my IDE shuffles them automatically. Anyway, its fixed now. Good catch.

File: pkg/webhook/validation.go
Comment (ocampeau) [reply pkg/webhook/validation.go]: I think `Found both set` is good. It makes sens to me at least.

**Comment by inteon** [comment]
/lgtm
Thanks a lot for the feature.
After this gets merged, we can do the v0.8.0 release. Also, @ocampeau could you create a documentation PR on the website to document this new feature?

**Comment by ocampeau** [comment]
Sure I can add documentation. I'll take a look into it

**Comment by ocampeau** [comment]
Documentation added with this PR: https://github.com/cert-manager/website/pull/1391

**Pull request merged**
